src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SerializableBytes.cs

```diff
@@ -337,25 +318,32 @@ private void EnsureCapacity(long value)
         }
 
         public override void SetLength(long value)
+            => SetLength(value, truncate: true);
+
+        public void SetLength(long value, bool truncate)
         {
             EnsureCapacity(value);
 
             if (value < length)
             {
-                // truncate the stream
+                if (truncate)
+                {
+                    var chunkIndex = GetChunkIndex(value);
+                    var chunkOffset = GetChunkOffset(value);
 
-                var chunkIndex = GetChunkIndex(value);
-                var chunkOffset = GetChunkOffset(value);
+                    Array.Clear(chunks[chunkIndex], chunkOffset, chunks[chunkIndex].Length - chunkOffset);
 
-                Array.Clear(chunks[chunkIndex], chunkOffset, chunks[chunkIndex].Length - chunkOffset);
+                    var trimIndex = chunkIndex + 1;
+                    for (var i = trimIndex; i < chunks.Count; i++)
+                        SharedPools.ByteArray.Free(chunks[i]);
 
-                var trimIndex = chunkIndex + 1;
-                for (var i = trimIndex; i < chunks.Count; i++)
+                    chunks.RemoveRange(trimIndex, chunks.Count - trimIndex);
+                }
+                else
                 {
-                    SharedPools.ByteArray.Free(chunks[i]);
+                    // TODO: do we want to clear out the remaining arrays?  Or is it fine to keep the existing data in
+                    // it just as garbage?
```

Comment:
we have pooled streams we use when writing data out to a temp-buffer before writing it out to the pipe.  We would prefer to not have that stream dump its internal byte[] "chunks" as those will just have to be reallocated on the next object we write out.  So that's why i made 'truncate' an option a client can opt out of.

Reply:
I'd vote to remove the TODO and just switch this to a comment indicating that we don't clear out the arrays when not truncating.

Reply:
done.